bgpv2: setting gobgp configuration based on new BGP APIs

harsimran-pabla opened the pull request:
Summary of changes 
- First commit: Adding default method in BGPv2 APIs  to set BGP peer default settings. ( similar to neighbor defaults [here](https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go#L323) )
- Second commit : This commit moves peer specific methods to separate file in gobgp pkg.
- Third commit : Modifying gobgp add/remove/update neighbor wrappers to use new APIs when possible. 
- Forth commit : Marking deprecated fields and removing unused VR field from NeighborRequest.

harsimran-pabla committed:
bgpv2: add default helpers for BGP peer configuration

Adding SetDefaults() method to CiliumBGPPeerConfigSpec, it will
setup uninitialized structures with default BGP values for timer,
transport, etc.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: move gobgp peer apis and unit tests to separate file

Moving add, remove and update neighbor APIs to separate file as
server.go is getting quite large with lot of functionality. Also,
splitting unit test file and moving peer tests to peer_test.go.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: gobgp peer configuration using new BGP API structures

Change includes peer configuration using BGPv2 APIs. There are three
main changes

- Include CiliumBGPNodePeer and CiliumBGPPeerConfigSpec in
  NeighborRequest structure. When NeighborRequest is used with these
  values set, equivalent gobgp APIs will be created based on these
  settings.
- Refactored peer API code to reuse common parts between BGPv1 and BGPv2 requests.
- Introducing setting source port for peer transport from BGPv2
  configuration.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs commented on pkg/bgpv1/types/bgp.go:
```diff
@@ -44,6 +44,9 @@ type Path struct {
 	VR       *v2alpha1api.CiliumBGPVirtualRouter
+
+	Peer   *v2alpha1api.CiliumBGPNodePeer
+	Config *v2alpha1api.CiliumBGPPeerConfigSpec
```
maybe name this `PeerConfig` instead of just `Config`? All these fields are essentially some config :)

rastislavs commented on pkg/bgpv1/types/bgp.go:
```diff
@@ -44,6 +44,9 @@ type Path struct {
 type NeighborRequest struct {
 	Neighbor *v2alpha1api.CiliumBGPNeighbor
 	VR       *v2alpha1api.CiliumBGPVirtualRouter
```
Should we mark these two fields as deprecated, so that their use is discouraged in the future? Or at least mention with a comment which fields should be used for which version, and that probably configuring both `Neighbor` and `Peer` is not valid?

BTW, is the `VR` even used somewhere in the gobgp code ATM? To me it does not seem to be.

harsimran-pabla replied:
I removed VR field - did this in separate commit to keep things easy to see.

rastislavs reviewed (commented):
thanks, the logic LGTM, I have just some suggestions about the BGP API

harsimran-pabla committed:
bgpv2: cleanup VR field from NeighborRequest

Marking deprecated fields in NeighborRequest and removing unused field
VR from it.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs reviewed (approved):
LGTM, thanks!
